src/main/java/com/google/devtools/build/lib/remote/RemoteSpawnRunner.java

```diff
@@ -291,15 +300,26 @@ private SpawnResult downloadAndFinalizeSpawnResult(
       SpawnExecutionContext context,
       RemoteOutputsMode remoteOutputsMode)
       throws ExecException, IOException, InterruptedException {
+    // If one output of a spawn is a top level output then download all outputs. Spawns
+    // are typically structured in a way that either all or no outputs are top level and
+    // it's much simpler to implement under this assumption.
+    boolean spawnHasTopLevelOutputs = remoteOutputsMode.downloadToplevelOutputsOnly() &&
+        hasTopLevelOutputs(spawn.getOutputFiles(), topLevelOutputs);
     SpawnResult.Status actionStatus =
         actionResult.getExitCode() == 0 ? Status.SUCCESS : Status.NON_ZERO_EXIT;
+    RemoteOutputsMode effectiveOutputsMode;
     // In case the action failed, download all outputs. It might be helpful for debugging
     // and there is no point in injecting output metadata of a failed action.
-    RemoteOutputsMode effectiveOutputsStrategy =
-        actionStatus == Status.SUCCESS ? remoteOutputsMode : RemoteOutputsMode.ALL;
+    if (actionStatus != Status.SUCCESS || spawnHasTopLevelOutputs) {
+      effectiveOutputsMode = RemoteOutputsMode.ALL;
+    } else if (remoteOutputsMode == RemoteOutputsMode.TOPLEVEL) {
+      effectiveOutputsMode = RemoteOutputsMode.MINIMAL;
```

Comment:
I'm a little confused here. What is the difference between `spawnHasTopLevelOutputs` and `remoteOutputsMode == RemoteOutputsMode.TOPLEVEL` cases?